File: implementations/styx-swift/Sources/Styx/Sexp.swift
/// S-expression formatting for compliance testing.
extension Document {
    public func toSexp() -> String {
        if entries.isEmpty {
            return "(document [-1, -1]\n)"
        }
        let entriesStr = entries.map { $0.toSexp(indent: 1) }.joined(separator: "\n")
        return "(document [-1, -1]\n\(entriesStr)\n)"
    }
}

extension Entry {
    func toSexp(indent: Int) -> String {
        let prefix = String(repeating: "  ", count: indent)
        let keyStr = key.toSexp(indent: indent + 1)
        let valueStr = value.toSexp(indent: indent + 1)
        return "\(prefix)(entry\n\(prefix)  \(keyStr)\n\(prefix)  \(valueStr))"
    }
}

extension Value {
    func toSexp(indent: Int) -> String {
        let prefix = String(repeating: "  ", count: indent)

        // Unit value (no tag, no payload)
        if tag == nil && payload == .none {
            return "(unit [\(span.start), \(span.end)])"
        }

        // Tag only (no payload)
        if let tag = tag, payload == .none {
            return "(tag [\(span.start), \(span.end)] \"\(tag.name)\")"
        }

        // Tag with payload
        if let tag = tag {
            let payloadStr = payloadToSexp(indent: indent + 1)
            return "(tag [\(span.start), \(span.end)] \"\(tag.name)\"\n\(prefix)  \(payloadStr))"
        }

        // Just payload
        return payloadToSexp(indent: indent)
    }

    private func payloadToSexp(indent: Int) -> String {
        let prefix = String(repeating: "  ", count: indent)

        switch payload {
        case .none:
            return "(unit [\(span.start), \(span.end)])"

        case .scalar(let scalar):
            let escaped = escapeString(scalar.text)
            return "(scalar [\(scalar.span.start), \(scalar.span.end)] \(scalar.kind.rawValue) \"\(escaped)\")"

        case .sequence(let seq):
            if seq.items.isEmpty {
                return "(sequence [\(seq.span.start), \(seq.span.end)])"
            }
            let items = seq.items.map { "\(prefix)  \($0.toSexp(indent: indent + 1))" }.joined(separator: "\n")
            return "(sequence [\(seq.span.start), \(seq.span.end)]\n\(items))"

        case .object(let obj):
            if obj.entries.isEmpty {
                return "(object [\(obj.span.start), \(obj.span.end)] \(obj.separator.rawValue))"
            }
            let entries = obj.entries.map { $0.toSexp(indent: indent + 1) }.joined(separator: "\n")
            return "(object [\(obj.span.start), \(obj.span.end)] \(obj.separator.rawValue)\n\(entries)\n\(prefix))"
        }
    }
}

func escapeString(_ s: String) -> String {
    var result = ""
    for char in s {
        switch char {
        case "\\": result += "\\\\"
        case "\"": result += "\\\""
        case "\n": result += "\\n"
        case "\r": result += "\\r"
        case "\t": result += "\\t"
        default: result.append(char)
        }
    }
    return result
}

extension ParseError {
    public func toSexp() -> String {
        let escaped = escapeString(message)
        return "(error [\(span.start), \(span.end)] \"parse error at \(span.start)-\(span.end): \(escaped)\")"
    }
}

File: implementations/styx-js/src/sexp.ts
import { Value, Scalar, Sequence, StyxObject, Entry, Document, Span } from "./types.js";

function jsonEscape(s: string): string {
  let result = "";
  for (const ch of s) {
    switch (ch) {
      case '"':
        result += '\\"';
        break;
      case "\\":
        result += "\\\\";
        break;
      case "\n":
        result += "\\n";
        break;
      case "\r":
        result += "\\r";
        break;
      case "\t":
        result += "\\t";
        break;
      default:
        if (ch.charCodeAt(0) < 32) {
          result += `\\u${ch.charCodeAt(0).toString(16).padStart(4, "0")}`;
        } else {
          result += ch;
        }
    }
  }
  return result;
}

function spanStr(span: Span): string {
  return `[${span.start}, ${span.end}]`;
}

function indent(level: number): string {
  return "  ".repeat(level);
}

export function documentToSexp(doc: Document): string {
  const lines: string[] = [];
  lines.push(`(document [-1, -1]`);

  for (const entry of doc.entries) {
    lines.push(entryToSexp(entry, 1));
  }

  lines.push(")");
  return lines.join("\n");
}

function entryToSexp(entry: Entry, level: number): string {
  const pad = indent(level);
  const lines: string[] = [];
  lines.push(`${pad}(entry`);
  lines.push(valueToSexp(entry.key, level + 1));
  lines.push(valueToSexp(entry.value, level + 1) + ")");
  return lines.join("\n");
}

function valueToSexp(value: Value, level: number): string {
  const pad = indent(level);
  const span = spanStr(value.span);

  // Unit value
  if (!value.tag && !value.payload) {
    return `${pad}(unit ${span})`;
  }

  // Tagged value
  if (value.tag) {
    const tagName = jsonEscape(value.tag.name);
    if (value.payload) {
      const payloadSexp = payloadToSexp(value.payload, level + 1);
      return `${pad}(tag ${span} "${tagName}"\n${payloadSexp})`;
    } else {
      return `${pad}(tag ${span} "${tagName}")`;
    }
  }

  // Untagged value with payload
  if (value.payload) {
    return payloadToSexp(value.payload, level);
  }

  return `${pad}(unit ${span})`;
}

function payloadToSexp(payload: Scalar | Sequence | StyxObject, level: number): string {
  const pad = indent(level);

  if (payload.type === "scalar") {
    const span = spanStr(payload.span);
    const text = jsonEscape(payload.text);
    return `${pad}(scalar ${span} ${payload.kind} "${text}")`;
  }

  if (payload.type === "sequence") {
    const span = spanStr(payload.span);
    if (payload.items.length === 0) {
      return `${pad}(sequence ${span})`;
    }
    const lines: string[] = [];
    lines.push(`${pad}(sequence ${span}`);
    for (const item of payload.items) {
      lines.push(valueToSexp(item, level + 1));
    }
    lines.push(`${pad.slice(2)})`);
    // Remove extra closing on last item and close here
    const result = lines.join("\n");
    // Fix: close sequence properly
    return result.replace(/\n\s*\)$/, ")");
  }

  if (payload.type === "object") {
    const span = spanStr(payload.span);
    if (payload.entries.length === 0) {
      return `${pad}(object ${span})`;
    }
    const lines: string[] = [];
    lines.push(`${pad}(object ${span}`);
    for (const entry of payload.entries) {
      lines.push(entryToSexp(entry, level + 1));
    }
    lines.push(`${pad})`);
    return lines.join("\n");
  }

  throw new Error(`Unknown payload type`);
}

export function errorToSexp(message: string, span: Span): string {
  const fullMsg = `parse error at ${span.start}-${span.end}: ${message}`;
  const msg = jsonEscape(fullMsg);
  return `(error [${span.start}, ${span.end}] "${msg}")`;
}
